[24.05] backport fcgiwrap instances fix for local privilege escalation issue

pacien opened the pull request:
## Description of changes

Backport of https://github.com/NixOS/nixpkgs/pull/318599.

Both `services.fcgiwrap.*` and the replacement options `services.fcgiwrap.instances.*` will co-exist on stable, with warnings when the former are used to notify users of the security issues, giving them time to migrate.

The consumer modules have been internally migrated too. For the cgit module, the user and group options were backported while keeping the previous implied default and showing a warning, allowing users to fix the situation gradually.


## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).


---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

pacien committed:
nixos/fcgiwrap: add deprecation notice and security warning

This deprecates the use of the global shared instance of fcgiwrap,
due to its security issues (running as root by default, actually
insecure control socket, allowing local remote escalation privileges,
with no fix due to the multiple consumers).

A warning is added to encourage users to migrate to properly isolated
instances (`services.fcgiwrap.instances.*`).

pacien committed:
nixos/cgit: use isolated fcgiwrap instance, add user/group options

This adds options to set the users and groups as which cgit instances
run, allowing the use of an unprivileged user instead of root.

"root" is kept as the default user to avoid breaking existing setups,
but a warning is shown in that case to alert the user.

Backport of:

commit 4f2da6c9c17d75ba43fbe85d5243a57735a5e4eb
    nixos/fcgiwrap: add option migration instruction errors
    (partial: move to instances)
commit 3d10deb7a5631e057bb5d84b5a6bd8fdf361a00a
    nixos/cgit: fix GIT_PROJECT_ROOT ownership
commit 2d8626bf0a35659a480c1a92bbd2682625a66e0f
    nixos/cgit: configurable user instead of root
commit c5dc3e203410bc3bfc77182cd8c6955b1bd64cfd
    nixos/fcgiwrap: adapt consumer modules and tests
commit 8101ae41f8cefce9e518a550881302c4f58a8c5b
    nixos/fcgiwrap: adapt consumer modules and tests
commit bf2ad6f48c95eea96768cc62dda7c6eb2097cbf4
    nixos/fcgiwrap: adapt consumer modules and tests

pacien committed:
nixos/zoneminder: use isolated fcgiwrap instance

Backport of:

commit fcb2a4a5fff1903d403955a9753a34f79bb24455
    nixos/zoneminder: set fcgiwrap socket owner
commit 4f2da6c9c17d75ba43fbe85d5243a57735a5e4eb
    nixos/fcgiwrap: add option migration instruction errors
    (partial: move to instances)
commit 8101ae41f8cefce9e518a550881302c4f58a8c5b
    nixos/fcgiwrap: adapt consumer modules and tests
commit bf2ad6f48c95eea96768cc62dda7c6eb2097cbf4
    nixos/fcgiwrap: adapt consumer modules and tests

pacien committed:
nixos/smokeping: use isolated fcgiwrap instance

This makes the CGI part of smokeping run as the unprivileged
"smokeping" user like the rest of the service (instead of root).

This also sets proper permissions for the fcgiwrap control socket.

Backport of:

commit 4f2da6c9c17d75ba43fbe85d5243a57735a5e4eb
    nixos/fcgiwrap: add option migration instruction errors
    (partial: move to instances)
commit c5dc3e203410bc3bfc77182cd8c6955b1bd64cfd
    nixos/fcgiwrap: adapt consumer modules and tests
commit 8101ae41f8cefce9e518a550881302c4f58a8c5b
    nixos/fcgiwrap: adapt consumer modules and tests
commit bf2ad6f48c95eea96768cc62dda7c6eb2097cbf4
    nixos/fcgiwrap: adapt consumer modules and tests

pacien committed:
nixos/fcgiwrap-instances: backport isolated multi-instance module

This backports the options `services.fcgiwrap.instances.*`,
allowing to configure isolated instances of fcgiwrap,
as an alternative to the global shared one.
This prepares the deprecation of the latter.

Backport of:

commit efc7aebda7f85f67b52cc334066c6dd344371103
    nixos/fcgiwrap: require explicit owner for UNIX sockets
commit 4f2da6c9c17d75ba43fbe85d5243a57735a5e4eb
    nixos/fcgiwrap: add option migration instruction errors
    (partial: move to instances)
commit 51b246a1acd4ce4926b8a78e3e7e0e6927d546ff
    nixos/fcgiwrap: do not run as root by default
commit 81f72015f0b96b1227a2de38409049fba0e73aad
    nixos/fcgiwrap: add unix socket owner, private by default
commit 289c1585c2a1f9ff9e159cbcdab664620dc9f7b3
    nixos/fcgiwrap: limit prefork type to positives
commit 3955eaf45015c9dd8a5a59412bf9c5e47b789a65
    nixos/fcgiwrap: improve readability of CLI args
commit 022289f2fadb3a3bad83273cd45d8a3e4753991e
    nixos/fcgiwrap: group options logically, fix doc
commit 41419ca2883f7a3294711faf4961d043868e27ef
    nixos/fcgiwrap: refactor for multiple instances

emilazy commented on nixos/modules/services/networking/cgit.nix:
```diff
@@ -165,29 +175,66 @@ in
+    warnings = flatten (flip mapAttrsToList cfgs (inst: cfg:
+      optional (cfg.user == "root") ''
+        `services.cgit.${inst}` is configured to run as root.
+        This has security implications: <TODO: advisory link>.
```
Thanks for your work on this. I think linking to the advisory is great; we should certainly post it on Discourse and arguably https://github.com/NixOS/nixpkgs/security/advisories (as‐yet unused).

I personally feel that a warning is not sufficient and that insecure configurations should be a hard error, because otherwise it’s likely they won’t be surfaced through automatic update pipelines etc.; however I realize there is a compatibility cost here.

minijackson replied:
> I personally feel that a warning is not sufficient and that insecure configurations should be a hard error, because otherwise it’s likely they won’t be surfaced through automatic update pipelines etc.; however I realize there is a compatibility cost here.

I think the biggest issue with that, especially on the stable channel, is that for people using the auto-upgrade and not really monitoring their machine, it would block other compatible security upgrades. Like for example the recent curl CVE.

emilazy replied:
There’s lots of circumstances where that would bite you. Packages *do* break unexpectedly. You need monitoring of failures to have security; Nix warnings, on the other hand, are rarely security‐relevant, and are trickier to automatically detect and monitor.

minijackson replied:
> There’s lots of circumstances where that would bite you. Packages _do_ break unexpectedly. You need monitoring of failures to have security; Nix warnings, on the other hand, are rarely security‐relevant, and are trickier to automatically detect and monitor.

I agree, you definitely should monitor your machines. But I think lots of people still don't.

But it's true that warnings would be harder to detect, hmm…

emilazy replied:
Perhaps we could emit a warning *and make the systemd services fail to start at runtime* with the same error. That way we fail closed in terms of this particular security vulnerability, anyone with any kind of monitoring (even manual) of “does the stuff I use even work at all” will notice that something’s amiss, and other security updates won’t be blocked.

minijackson replied:
Hmmm, interesting. But note that this might be more severe than an eval failure, since in the case of an eval failure, the previously built system still continues running. What about making another dummy systemd service that fails? That way at least they monitor failing services, they'll be notified, without breaking anything.

minijackson replied:
There really should be a proper notification channel for these kinds of things…

emilazy replied:
Well, the system would continue running, just not the part of it with a privilege escalation vulnerability…

Fundamentally we have to do *something* that people will notice to get their attention about a security vulnerability if we don’t want to leave many hosts vulnerable to it.

minijackson replied:
Yeah, which is why I think having a proper security ML or something would be good for these exact cases.

Right now we have to choose whether removing the vulnerability is more important than keeping the service running. In this instance, one of the affected services is zoneminder, which can be used to make automatic video-surveillance. Is a local privilege escalation more important than the user's video-surveillance?

One thing I'm sure is that I don't like being in that position :sweat_smile:

emilazy replied:
Well, I’d still be just fine with failing the build if I’ve talked you into it!

Seriously though, a warning is better than nothing. But I do think we can do better.

By the way, you can subscribe to email updates from the [Announcements → Security](https://discourse.nixos.org/c/announcements/security/56) category on the Discourse; that’s effectively the advisories mailing list. Discoverability is another thing, though, as is making sure everything important actually gets posted there.

bendlas replied:
> Well, I’d still be just fine with failing the build if I’ve talked you into it!

👎👎👎

If you feel that strongly about it, why not invest that energy into giving better visibility to warnings? E.g. by having autoUpdate send emails?

There are literally thousands of local privilege escalations on any given linux. Why not hunt for those instead of harrassing stable users?

emilazy replied:
A warnings/notification system that actually works is an important thing that we’re missing across the entirety of Nixpkgs and NixOS. It would be a big improvement, and there has already been some work on similar things in terms of the problems infrastructure for packages. However, it would take a lot more energy and time than giving advice on this issue.

I think “literally thousands” is an exaggeration. Linux platform security is not in a great state, but it is not *that* bad, especially in a server environment. The existence of other vulnerabilities shouldn’t stop us correctly handling this one.

I reject the idea that letting someone know their system lets *any local user* escalate to root constitutes “harassment”. I think you have made your position very clear by now, however.

emilazy replied:
Throwing an evaluation error is pretty much exactly the same behaviour as marking packages with `knownVulnerabilities`, which is a normal thing we do all the time to let people know that they are depending on vulnerable software, including in the stable branch. If we want to decide we should do something more nuanced, then it should address the far more common case of `knownVulnerabilities` first. In the meantime, we already have established precedent that we block builds of known‐insecure configurations by default and that you need to monitor failed updates to ensure you get updates (as of course you would have to anyway). This does not impact availability because failing to upgrade the system does not take down the existing one.

emilazy replied:
I think we could just add a `services.fcgiwrap.allowLocalPrivilegeEscalationUntilNextStableRelease` option or something if we want an escape hatch.

I agree that the systemd service failure option is not ideal.

The GitHub advisory tab lets us assign CVEs, as I understand it, so it would actually be quite nice to use. But only repository admins can post stuff; we’d probably need to organize procedures and someone to be responsible for it.

LeSuisse replied:
GHA are not really adequate in our situation with the current way the permissions are handled. Creation and publication needs to be handled by repo admins and cannot be delegated :/ (also not sure if tying more things with the GH ecosystem is a good play here).

In this specific case at this point the cat is out of the bag so I suggest to go ahead and publish the advisory on https://discourse.nixos.org/c/announcements/security/56 when this is ready to be merged to get the final link so it can be added to the messages.

emilazy replied:
I think Discourse will be okay, but there’s indeed a bit of a race condition there. Maybe we can post something set to private and then let it go public once we merge? I’m not sure who has that kind of access on Discourse other than the moderation team.

bendlas reviewed (commented):
PR looks clean and sensible to me. Would approve.

About the advisory: https://gist.github.com/pacien/82119f8e1a6569c5b12bf0f75ecc3b9a

> Cache files owned by root may however need to be manually cleared.

What's the impact of having root-owned cache files left over? Does anything break?
Would it make sense to have a pre-start script to either clear cache files of the wrong user, or chown them?

Overall thanks for taking this on!

bendlas commented:
> I just added a `chown` in the existing `preStart` script, so that gets taken care of automatically.

Do you think that's something that should go to unstable? I could help with forward-porting that ...

bendlas commented on nixos/modules/services/networking/smokeping.nix:
```diff
@@ -337,7 +338,11 @@ in
-    services.fcgiwrap.enable = mkIf cfg.webService true;
+    services.fcgiwrap.instances.smokeping = mkIf cfg.webService {
+      process.user = cfg.user;
+      process.group = cfg.user;
```
```suggestion
      process.group = cfg.group;
```

bendlas replied:
err .. just seeing that that was intentional, disregard

bendlas commented on nixos/modules/services/networking/smokeping.nix:
```diff
@@ -328,6 +328,7 @@ in
       };
       preStart = ''
         mkdir -m 0755 -p ${smokepingHome}/cache ${smokepingHome}/data
+        chown -R ${cfg.user}:${cfg.user} ${smokepingHome}/{cache,data}
```
```suggestion
        chown -R ${cfg.user}:${cfg.group} ${smokepingHome}/{cache,data}
```

bendlas replied:
err .. just seeing that that was intentional, disregard

bendlas reviewed (changes requested):
Seems like we're using `using` instead of `group` in a couple of places, which incidentally works in the default config where `user == group`

**EDIT** err .. just seeing that that was intentional, disregard

bendlas commented:
All good points. I'm halfway through putting up a PR to main. I did go the route of managing the current directory structure with `systemd.tmpfiles`, but `/var/cache` et al, managed directly by the service file is probably an even better idea.

pacien committed:
nixos/fcgiwrap: fail eval with security assertion

This adds a security assertion when using the global instance of
fcgiwrap, which is vulnerable to a local privilege escalation.

This is in addition to the current evaluation warning, and is more in
line with being loud with security issues, similarly to with vulnerable
packages.

The evaluation failure can nevertheless be bypassed by setting:
`services.fcgiwrap.allowGlobalInstanceLocalPrivilegeEscalation = true`.

emilazy reviewed (approved):
This PR looks good to me in its current state; sorry for leaving this hanging for so long. The advisory should maybe be updated to mention `services.fcgiwrap.allowGlobalInstanceLocalPrivilegeEscalation`. Tagging @LeSuisse for a final review.

There is an awkward timing thing here: we’d ideally post the advisory publicly at the point where this is merged and users on 24.05 can actually follow the mitigation, but that would mean that we couldn’t include the advisory link in the corresponding channel bump. Users can subscribe to the security announcements category on Discourse and get emailed alerts, so timing does matter. I think notifications will work fine because every post in that category needs to be approved anyway. I think that making the post private and including the URL here, then later making it public when the PR hits the channels, is probably the best approach. cc @mweinelt who has the perms to approve such posts

emilazy commented:
Another option would be to make the advisory public early, but clarify that users on 24.05 will need to wait for the PR to hit the channels and maybe link one of the trackers for that. That way we don’t need to do any special coordination of the timing.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/security-advisory-local-privilege-escalation-in-the-fcgiwrap-nixos-module-also-affecting-the-cgit-smokeping-and-zoneminder-modules/51419/1

emilazy commented:
There was a slight communication mishap and it’s been published already 😅

That should be fine though since it already points to the PR trackers, so if you edit in the link we can finally get this merged. Thank you for all your work on this issue 💜

pacien committed:
nixos/fcgiwrap: add security advisory links to messages
